// @databricks/sql/dist/connection/auth/DatabricksOAuth/OAuthManager.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureOAuthManager = exports.AWSOAuthManager = void 0;
const openid_client_1 = require("openid-client");
const HiveDriverError_1 = __importDefault(require("../../../errors/HiveDriverError"));
const IDBSQLLogger_1 = require("../../../contracts/IDBSQLLogger");
const OAuthToken_1 = __importDefault(require("./OAuthToken"));
const AuthorizationCode_1 = __importDefault(require("./AuthorizationCode"));
const OAuthScope_1 = require("./OAuthScope");
function getDatabricksOIDCUrl(host) {
    const schema = host.startsWith('https://') ? '' : 'https://';
    const trailingSlash = host.endsWith('/') ? '' : '/';
    return `${schema}${host}${trailingSlash}oidc`;
}
class OAuthManager {
    constructor(options) {
        this.options = options;
        this.context = options.context;
    }
    getScopes(requestedScopes) {
        return requestedScopes;
    }
    getClient() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            // Obtain http agent each time when we need an OAuth client
            // to ensure that we always use a valid agent instance
            const connectionProvider = yield this.context.getConnectionProvider();
            this.agent = yield connectionProvider.getAgent();
            const getHttpOptions = () => ({
                agent: this.agent,
            });
            if (!this.issuer) {
                // To use custom http agent in Issuer.discover(), we'd have to set Issuer[custom.http_options].
                // However, that's a static field, and if multiple instances of OAuthManager used, race condition
                // may occur when they simultaneously override that field and then try to use Issuer.discover().
                // Therefore we create a local class derived from Issuer, and set that field for it, thus making
                // sure that it will not interfere with other instances (or other code that may use Issuer)
                class CustomIssuer extends openid_client_1.Issuer {
                }
                _a = openid_client_1.custom.http_options;
                CustomIssuer[_a] = getHttpOptions;
                const issuer = yield CustomIssuer.discover(this.getOIDCConfigUrl());
                // Overwrite `authorization_endpoint` in default config (specifically needed for Azure flow
                // where this URL has to be different)
                this.issuer = new openid_client_1.Issuer(Object.assign(Object.assign({}, issuer.metadata), { authorization_endpoint: this.getAuthorizationUrl() }));
                this.issuer[openid_client_1.custom.http_options] = getHttpOptions;
            }
            if (!this.client) {
                this.client = new this.issuer.Client({
                    client_id: this.getClientId(),
                    client_secret: this.options.clientSecret,
                    token_endpoint_auth_method: this.options.clientSecret === undefined ? 'none' : 'client_secret_basic',
                });
                this.client[openid_client_1.custom.http_options] = getHttpOptions;
            }
            return this.client;
        });
    }
    refreshAccessTokenU2M(token) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!token.refreshToken) {
                const message = `OAuth access token expired on ${token.expirationTime}.`;
                this.context.getLogger().log(IDBSQLLogger_1.LogLevel.error, message);
                throw new HiveDriverError_1.default(message);
            }
            // Try to refresh using the refresh token
            this.context
                .getLogger()
                .log(IDBSQLLogger_1.LogLevel.debug, `Attempting to refresh OAuth access token that expired on ${token.expirationTime}`);
            const client = yield this.getClient();
            const { access_token: accessToken, refresh_token: refreshToken } = yield client.refresh(token.refreshToken);
            if (!accessToken || !refreshToken) {
                throw new Error('Failed to refresh token: invalid response');
            }
            return new OAuthToken_1.default(accessToken, refreshToken);
        });
    }
    refreshAccessTokenM2M() {
        return __awaiter(this, void 0, void 0, function* () {
            const { access_token: accessToken, refresh_token: refreshToken } = yield this.getTokenM2M();
            if (!accessToken) {
                throw new Error('Failed to fetch access token');
            }
            return new OAuthToken_1.default(accessToken, refreshToken);
        });
    }
    refreshAccessToken(token) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                if (!token.hasExpired) {
                    // The access token is fine. Just return it.
                    return token;
                }
            }
            catch (error) {
                this.context.getLogger().log(IDBSQLLogger_1.LogLevel.error, `${error}`);
                throw error;
            }
            return this.options.clientSecret === undefined ? this.refreshAccessTokenU2M(token) : this.refreshAccessTokenM2M();
        });
    }
    getTokenU2M(scopes) {
        return __awaiter(this, void 0, void 0, function* () {
            const client = yield this.getClient();
            const authCode = new AuthorizationCode_1.default({
                client,
                ports: this.getCallbackPorts(),
                context: this.context,
            });
            const mappedScopes = this.getScopes(scopes);
            const { code, verifier, redirectUri } = yield authCode.fetch(mappedScopes);
            return client.grant({
                grant_type: 'authorization_code',
                code,
                code_verifier: verifier,
                redirect_uri: redirectUri,
            });
        });
    }
    getTokenM2M() {
        return __awaiter(this, void 0, void 0, function* () {
            const client = yield this.getClient();
            // M2M flow doesn't really support token refreshing, and refresh should not be available
            // in response. Each time access token expires, client can just acquire a new one using
            // client secret. Here we explicitly return access token only as a sign that we're not going
            // to use refresh token for M2M flow anywhere later
            const { access_token: accessToken } = yield client.grant({
                grant_type: 'client_credentials',
                scope: 'all-apis', // this is the only allowed scope for M2M flow
            });
            return { access_token: accessToken, refresh_token: undefined };
        });
    }
    getToken(scopes) {
        return __awaiter(this, void 0, void 0, function* () {
            const { access_token: accessToken, refresh_token: refreshToken } = this.options.clientSecret === undefined ? yield this.getTokenU2M(scopes) : yield this.getTokenM2M();
            if (!accessToken) {
                throw new Error('Failed to fetch access token');
            }
            return new OAuthToken_1.default(accessToken, refreshToken);
        });
    }
    static getManager(options) {
        // normalize
        const host = options.host.toLowerCase().replace('https://', '').split('/')[0];
        // eslint-disable-next-line @typescript-eslint/no-use-before-define
        const managers = [AWSOAuthManager, AzureOAuthManager];
        for (const OAuthManagerClass of managers) {
            for (const domain of OAuthManagerClass.domains) {
                if (host.endsWith(domain)) {
                    return new OAuthManagerClass(options);
                }
            }
        }
        throw new Error(`OAuth is not supported for ${options.host}`);
    }
}
exports.default = OAuthManager;
class AWSOAuthManager extends OAuthManager {
    getOIDCConfigUrl() {
        return `${getDatabricksOIDCUrl(this.options.host)}/.well-known/oauth-authorization-server`;
    }
    getAuthorizationUrl() {
        return `${getDatabricksOIDCUrl(this.options.host)}/oauth2/v2.0/authorize`;
    }
    getClientId() {
        var _a;
        return (_a = this.options.clientId) !== null && _a !== void 0 ? _a : AWSOAuthManager.defaultClientId;
    }
    getCallbackPorts() {
        var _a;
        return (_a = this.options.callbackPorts) !== null && _a !== void 0 ? _a : AWSOAuthManager.defaultCallbackPorts;
    }
}
exports.AWSOAuthManager = AWSOAuthManager;
AWSOAuthManager.domains = ['.cloud.databricks.com', '.dev.databricks.com'];
AWSOAuthManager.defaultClientId = 'databricks-sql-connector';
AWSOAuthManager.defaultCallbackPorts = [8030];
class AzureOAuthManager extends OAuthManager {
    getOIDCConfigUrl() {
        return 'https://login.microsoftonline.com/organizations/v2.0/.well-known/openid-configuration';
    }
    getAuthorizationUrl() {
        return `${getDatabricksOIDCUrl(this.options.host)}/oauth2/v2.0/authorize`;
    }
    getClientId() {
        var _a;
        return (_a = this.options.clientId) !== null && _a !== void 0 ? _a : AzureOAuthManager.defaultClientId;
    }
    getCallbackPorts() {
        var _a;
        return (_a = this.options.callbackPorts) !== null && _a !== void 0 ? _a : AzureOAuthManager.defaultCallbackPorts;
    }
    getScopes(requestedScopes) {
        var _a;
        // There is no corresponding scopes in Azure, instead, access control will be delegated to Databricks
        const tenantId = (_a = this.options.azureTenantId) !== null && _a !== void 0 ? _a : AzureOAuthManager.datatricksAzureApp;
        const azureScopes = [`${tenantId}/user_impersonation`];
        if (requestedScopes.includes(OAuthScope_1.OAuthScope.offlineAccess)) {
            azureScopes.push(OAuthScope_1.OAuthScope.offlineAccess);
        }
        return azureScopes;
    }
}
exports.AzureOAuthManager = AzureOAuthManager;
AzureOAuthManager.domains = ['.azuredatabricks.net', '.databricks.azure.cn', '.databricks.azure.us'];
AzureOAuthManager.defaultClientId = '96eecda7-19ea-49cc-abb5-240097d554f5';
AzureOAuthManager.defaultCallbackPorts = [8030];
AzureOAuthManager.datatricksAzureApp = '2ff814a6-3304-4ab8-85cb-cd0e6f879c1d';
