// @databricks/sql/dist/result/CloudFetchResult.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const buffer_1 = require("buffer");
const node_fetch_1 = __importDefault(require("node-fetch"));
const ArrowResult_1 = __importDefault(require("./ArrowResult"));
const globalConfig_1 = __importDefault(require("../globalConfig"));
class CloudFetchResult extends ArrowResult_1.default {
    constructor(context, schema) {
        // Arrow schema returned in metadata is not needed for CloudFetch results:
        // each batch already contains schema and could be decoded as is
        super(context, schema, buffer_1.Buffer.alloc(0));
        this.pendingLinks = [];
        this.downloadedBatches = [];
    }
    hasPendingData() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.pendingLinks.length > 0 || this.downloadedBatches.length > 0;
        });
    }
    getBatches(data) {
        return __awaiter(this, void 0, void 0, function* () {
            data.forEach((item) => {
                var _a;
                (_a = item.resultLinks) === null || _a === void 0 ? void 0 : _a.forEach((link) => {
                    this.pendingLinks.push(link);
                });
            });
            if (this.downloadedBatches.length === 0) {
                const links = this.pendingLinks.splice(0, globalConfig_1.default.cloudFetchConcurrentDownloads);
                const tasks = links.map((link) => this.downloadLink(link));
                const batches = yield Promise.all(tasks);
                this.downloadedBatches.push(...batches);
            }
            return this.downloadedBatches.splice(0, 1);
        });
    }
    downloadLink(link) {
        return __awaiter(this, void 0, void 0, function* () {
            if (Date.now() >= link.expiryTime.toNumber()) {
                throw new Error('CloudFetch link has expired');
            }
            const response = yield this.fetch(link.fileLink);
            if (!response.ok) {
                throw new Error(`CloudFetch HTTP error ${response.status} ${response.statusText}`);
            }
            const result = yield response.arrayBuffer();
            return buffer_1.Buffer.from(result);
        });
    }
    fetch(url, init) {
        return __awaiter(this, void 0, void 0, function* () {
            const connectionProvider = yield this.context.getConnectionProvider();
            const agent = yield connectionProvider.getAgent();
            return (0, node_fetch_1.default)(url, Object.assign({ agent }, init));
        });
    }
}
exports.default = CloudFetchResult;
