// @databricks/sql/dist/DBSQLOperation/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const uuid_1 = require("uuid");
const TCLIService_types_1 = require("../../thrift/TCLIService_types");
const Status_1 = __importDefault(require("../dto/Status"));
const FetchResultsHelper_1 = __importDefault(require("./FetchResultsHelper"));
const IDBSQLLogger_1 = require("../contracts/IDBSQLLogger");
const OperationStateError_1 = __importStar(require("../errors/OperationStateError"));
const JsonResult_1 = __importDefault(require("../result/JsonResult"));
const ArrowResult_1 = __importDefault(require("../result/ArrowResult"));
const CloudFetchResult_1 = __importDefault(require("../result/CloudFetchResult"));
const utils_1 = require("../utils");
const HiveDriverError_1 = __importDefault(require("../errors/HiveDriverError"));
const defaultMaxRows = 100000;
function delay(ms) {
    return __awaiter(this, void 0, void 0, function* () {
        return new Promise((resolve) => {
            setTimeout(() => {
                resolve();
            }, ms);
        });
    });
}
class DBSQLOperation {
    constructor({ handle, directResults, context }) {
        this.closed = false;
        this.cancelled = false;
        this.state = TCLIService_types_1.TOperationState.INITIALIZED_STATE;
        this.hasResultSet = false;
        this.operationHandle = handle;
        this.context = context;
        const useOnlyPrefetchedResults = Boolean(directResults === null || directResults === void 0 ? void 0 : directResults.closeOperation);
        this.hasResultSet = this.operationHandle.hasResultSet;
        if (directResults === null || directResults === void 0 ? void 0 : directResults.operationStatus) {
            this.processOperationStatusResponse(directResults.operationStatus);
        }
        this.metadata = directResults === null || directResults === void 0 ? void 0 : directResults.resultSetMetadata;
        this._data = new FetchResultsHelper_1.default(this.context, this.operationHandle, [directResults === null || directResults === void 0 ? void 0 : directResults.resultSet], useOnlyPrefetchedResults);
        this.closeOperation = directResults === null || directResults === void 0 ? void 0 : directResults.closeOperation;
        this.context.getLogger().log(IDBSQLLogger_1.LogLevel.debug, `Operation created with id: ${this.getId()}`);
    }
    getId() {
        var _a, _b;
        return (0, uuid_1.stringify)(((_b = (_a = this.operationHandle) === null || _a === void 0 ? void 0 : _a.operationId) === null || _b === void 0 ? void 0 : _b.guid) || (0, uuid_1.parse)(uuid_1.NIL));
    }
    /**
     * Fetches all data
     * @public
     * @param options - maxRows property can be set to limit chunk size
     * @returns Array of data with length equal to option.maxRows
     * @throws {StatusError}
     * @example
     * const result = await queryOperation.fetchAll();
     */
    fetchAll(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const data = [];
            do {
                // eslint-disable-next-line no-await-in-loop
                const chunk = yield this.fetchChunk(options);
                data.push(chunk);
            } while (yield this.hasMoreRows()); // eslint-disable-line no-await-in-loop
            this.context.getLogger().log(IDBSQLLogger_1.LogLevel.debug, `Fetched all data from operation with id: ${this.getId()}`);
            return data.flat();
        });
    }
    /**
     * Fetches chunk of data
     * @public
     * @param options - maxRows property sets chunk size
     * @returns Array of data with length equal to option.maxRows
     * @throws {StatusError}
     * @example
     * const result = await queryOperation.fetchChunk({maxRows: 1000});
     */
    fetchChunk(options) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.failIfClosed();
            if (!this.hasResultSet) {
                return [];
            }
            yield this.waitUntilReady(options);
            const [resultHandler, data] = yield Promise.all([
                this.getResultHandler(),
                this._data.fetch((options === null || options === void 0 ? void 0 : options.maxRows) || defaultMaxRows),
            ]);
            yield this.failIfClosed();
            const result = yield resultHandler.getValue(data ? [data] : []);
            this.context
                .getLogger()
                .log(IDBSQLLogger_1.LogLevel.debug, `Fetched chunk of size: ${(options === null || options === void 0 ? void 0 : options.maxRows) || defaultMaxRows} from operation with id: ${this.getId()}`);
            return result;
        });
    }
    /**
     * Requests operation status
     * @param progress
     * @throws {StatusError}
     */
    status(progress = false) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.failIfClosed();
            this.context.getLogger().log(IDBSQLLogger_1.LogLevel.debug, `Fetching status for operation with id: ${this.getId()}`);
            if (this.operationStatus) {
                return this.operationStatus;
            }
            const driver = yield this.context.getDriver();
            const response = yield driver.getOperationStatus({
                operationHandle: this.operationHandle,
                getProgressUpdate: progress,
            });
            return this.processOperationStatusResponse(response);
        });
    }
    /**
     * Cancels operation
     * @throws {StatusError}
     */
    cancel() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (this.closed || this.cancelled) {
                return Status_1.default.success();
            }
            this.context.getLogger().log(IDBSQLLogger_1.LogLevel.debug, `Cancelling operation with id: ${this.getId()}`);
            const driver = yield this.context.getDriver();
            const response = yield driver.cancelOperation({
                operationHandle: this.operationHandle,
            });
            Status_1.default.assert(response.status);
            this.cancelled = true;
            const result = new Status_1.default(response.status);
            // Cancelled operation becomes unusable, similarly to being closed
            (_a = this.onClose) === null || _a === void 0 ? void 0 : _a.call(this);
            return result;
        });
    }
    /**
     * Closes operation
     * @throws {StatusError}
     */
    close() {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            if (this.closed || this.cancelled) {
                return Status_1.default.success();
            }
            this.context.getLogger().log(IDBSQLLogger_1.LogLevel.debug, `Closing operation with id: ${this.getId()}`);
            const driver = yield this.context.getDriver();
            const response = (_a = this.closeOperation) !== null && _a !== void 0 ? _a : (yield driver.closeOperation({
                operationHandle: this.operationHandle,
            }));
            Status_1.default.assert(response.status);
            this.closed = true;
            const result = new Status_1.default(response.status);
            (_b = this.onClose) === null || _b === void 0 ? void 0 : _b.call(this);
            return result;
        });
    }
    finished(options) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.failIfClosed();
            yield this.waitUntilReady(options);
        });
    }
    hasMoreRows() {
        return __awaiter(this, void 0, void 0, function* () {
            // If operation is closed or cancelled - we should not try to get data from it
            if (this.closed || this.cancelled) {
                return false;
            }
            // Return early if there are still data available for fetching
            if (this._data.hasMoreRows) {
                return true;
            }
            // If we fetched all the data from server - check if there's anything buffered in result handler
            const resultHandler = yield this.getResultHandler();
            return resultHandler.hasPendingData();
        });
    }
    getSchema(options) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            yield this.failIfClosed();
            if (!this.hasResultSet) {
                return null;
            }
            yield this.waitUntilReady(options);
            this.context.getLogger().log(IDBSQLLogger_1.LogLevel.debug, `Fetching schema for operation with id: ${this.getId()}`);
            const metadata = yield this.fetchMetadata();
            return (_a = metadata.schema) !== null && _a !== void 0 ? _a : null;
        });
    }
    getMetadata() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.failIfClosed();
            yield this.waitUntilReady();
            return this.fetchMetadata();
        });
    }
    failIfClosed() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.closed) {
                throw new OperationStateError_1.default(OperationStateError_1.OperationStateErrorCode.Closed);
            }
            if (this.cancelled) {
                throw new OperationStateError_1.default(OperationStateError_1.OperationStateErrorCode.Canceled);
            }
        });
    }
    waitUntilReady(options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.state === TCLIService_types_1.TOperationState.FINISHED_STATE) {
                return;
            }
            let isReady = false;
            while (!isReady) {
                // eslint-disable-next-line no-await-in-loop
                const response = yield this.status(Boolean(options === null || options === void 0 ? void 0 : options.progress));
                if (options === null || options === void 0 ? void 0 : options.callback) {
                    // eslint-disable-next-line no-await-in-loop
                    yield Promise.resolve(options.callback(response));
                }
                switch (response.operationState) {
                    // For these states do nothing and continue waiting
                    case TCLIService_types_1.TOperationState.INITIALIZED_STATE:
                    case TCLIService_types_1.TOperationState.PENDING_STATE:
                    case TCLIService_types_1.TOperationState.RUNNING_STATE:
                        break;
                    // Operation is completed, so exit the loop
                    case TCLIService_types_1.TOperationState.FINISHED_STATE:
                        isReady = true;
                        break;
                    // Operation was cancelled, so set a flag and exit the loop (throw an error)
                    case TCLIService_types_1.TOperationState.CANCELED_STATE:
                        this.cancelled = true;
                        throw new OperationStateError_1.default(OperationStateError_1.OperationStateErrorCode.Canceled, response);
                    // Operation was closed, so set a flag and exit the loop (throw an error)
                    case TCLIService_types_1.TOperationState.CLOSED_STATE:
                        this.closed = true;
                        throw new OperationStateError_1.default(OperationStateError_1.OperationStateErrorCode.Closed, response);
                    // Error states - throw and exit the loop
                    case TCLIService_types_1.TOperationState.ERROR_STATE:
                        throw new OperationStateError_1.default(OperationStateError_1.OperationStateErrorCode.Error, response);
                    case TCLIService_types_1.TOperationState.TIMEDOUT_STATE:
                        throw new OperationStateError_1.default(OperationStateError_1.OperationStateErrorCode.Timeout, response);
                    case TCLIService_types_1.TOperationState.UKNOWN_STATE:
                    default:
                        throw new OperationStateError_1.default(OperationStateError_1.OperationStateErrorCode.Unknown, response);
                }
                // If not ready yet - make some delay before the next status requests
                if (!isReady) {
                    // eslint-disable-next-line no-await-in-loop
                    yield delay(100);
                }
            }
        });
    }
    fetchMetadata() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.metadata) {
                const driver = yield this.context.getDriver();
                const metadata = yield driver.getResultSetMetadata({
                    operationHandle: this.operationHandle,
                });
                Status_1.default.assert(metadata.status);
                this.metadata = metadata;
            }
            return this.metadata;
        });
    }
    getResultHandler() {
        return __awaiter(this, void 0, void 0, function* () {
            const metadata = yield this.fetchMetadata();
            const resultFormat = (0, utils_1.definedOrError)(metadata.resultFormat);
            if (!this.resultHandler) {
                switch (resultFormat) {
                    case TCLIService_types_1.TSparkRowSetType.COLUMN_BASED_SET:
                        this.resultHandler = new JsonResult_1.default(this.context, metadata.schema);
                        break;
                    case TCLIService_types_1.TSparkRowSetType.ARROW_BASED_SET:
                        this.resultHandler = new ArrowResult_1.default(this.context, metadata.schema, metadata.arrowSchema);
                        break;
                    case TCLIService_types_1.TSparkRowSetType.URL_BASED_SET:
                        this.resultHandler = new CloudFetchResult_1.default(this.context, metadata.schema);
                        break;
                    default:
                        this.resultHandler = undefined;
                        break;
                }
            }
            if (!this.resultHandler) {
                throw new HiveDriverError_1.default(`Unsupported result format: ${TCLIService_types_1.TSparkRowSetType[resultFormat]}`);
            }
            return this.resultHandler;
        });
    }
    processOperationStatusResponse(response) {
        var _a;
        Status_1.default.assert(response.status);
        this.state = (_a = response.operationState) !== null && _a !== void 0 ? _a : this.state;
        if (typeof response.hasResultSet === 'boolean') {
            this.hasResultSet = response.hasResultSet;
        }
        const isInProgress = [
            TCLIService_types_1.TOperationState.INITIALIZED_STATE,
            TCLIService_types_1.TOperationState.PENDING_STATE,
            TCLIService_types_1.TOperationState.RUNNING_STATE,
        ].includes(this.state);
        if (!isInProgress) {
            this.operationStatus = response;
        }
        return response;
    }
}
exports.default = DBSQLOperation;
